src/exercise/02.js

```diff
@@ -11,73 +11,102 @@ import {
-  // here's how you use it:
-  // const state = useAsync(
-  //   () => {
-  //     if (!pokemonName) {
-  //       return
-  //     }
-  //     return fetchPokemon(pokemonName)
-  //   },
-  //   {status: pokemonName ? 'pending' : 'idle'},
-  //   [pokemonName],
-  // )
-  // 🐨 so your job is to create a useAsync function that makes this work.
-  const [state, dispatch] = React.useReducer(pokemonInfoReducer, {
-    status: pokemonName ? 'pending' : 'idle',
-    // 🐨 this will need to be "data" instead of "pokemon"
-    pokemon: null,
-    error: null,
-  })
+function useDispatchRef(dispatch) {
+  // Set up a ref to check lifecycle status.
+  const isMounted = React.useRef(false);
 
   React.useEffect(() => {
-    // 💰 this first early-exit bit is a little tricky, so let me give you a hint:
-    // const promise = asyncCallback()
-    // if (!promise) {
-    //   return
-    // }
-    // then you can dispatch and handle the promise etc...
-    if (!pokemonName) {
-      return
+    // we are mounted, we can run the dispatch stuff.
+    isMounted.current = true;
+
+    console.log('we are mounted');
+    return () => {
+      isMounted.current = false;
+      console.log('we are unmounted');
     }
+  });
+
+  // Memoize this so we don't end up with exhaustive deps error.
+  return React.useCallback((...args) => { // Pass in any args that the function needs.
+      isMounted.current ? dispatch(...args) : void 0; // If we are mounted, run it, otherwise void.
+    },
+    [dispatch],
+  );
+}
+
+function useAsync(initialState) {
+  // Get the updated state (and dispatch function);
+  const [state, reducerDispatch] = React.useReducer(asyncReducer, {
+    status: 'idle',
+    data: null,
+    error: null,
+    ...initialState,
+  });
+
+  const dispatch = useDispatchRef(reducerDispatch);
+
+  // Destructure state so we can return just what we need.
+  const {data, error, status} = state;
+
+  // Add this run method that can we control memoization of.
+  // useCallback makes this function consistent over time.
+  const runFunc = React.useCallback((promise) => {
     dispatch({type: 'pending'})
-    fetchPokemon(pokemonName).then(
-      pokemon => {
-        dispatch({type: 'resolved', pokemon})
+    promise.then(
+      data => {
+        dispatch({type: 'resolved', data})
       },
       error => {
         dispatch({type: 'rejected', error})
       },
-    )
-    // 🐨 you'll accept dependencies as an array and pass that here.
-    // 🐨 because of limitations with ESLint, you'll need to ignore
-    // the react-hooks/exhaustive-deps rule. We'll fix this in an extra credit.
-  }, [pokemonName])
-
-  // 🐨 this will change from "pokemon" to "data"
-  const {pokemon, status, error} = state
+    );
+  }, [dispatch]); // Our dispatch function comes from our reducer and will never change, so this is fine.
+
+  // Return the goods.
+  return {
+    error,
+    status,
+    data,
+    runFunc,
+  };
```

Comment:
Consider returning an array so the callback is distinguished from the state values. This would be a little more consistent with some of React's core hooks (`useState` comes to mind first and foremost. Example:


```suggestion
  return [
    { error, status, data },
    runFunc,
  ];
```